config/os: add support for clickable tilde paths

gsLocker opened the pull request:
Enables clicking on tilde-based paths like ~/.config/file.txt. When clicked, the path is expanded to the user's home directory and opened.

Related to #1972

This addresses the maintainer feedback from PR #5813 by reusing the existing os.expandHome() function rather than duplicating tilde expansion logic. The implementation is cross-platform (Linux/macOS/FreeBSD) from the start.

Implementation:
- Modified URL regex to detect ~/path patterns (excludes bare ~ and ~user)
- Removed :,. from file path character class to avoid underlining punctuation
- Added negative lookbehind (?<!\.) to prevent trailing period matches
- Reuse os.expandHome() for tilde expansion in os/open.zig
- Added safety trimming for edge case punctuation handling
- Added test cases for positive and negative scenarios

The regex matches ~/path but not bare ~ or ~user/path (which would require different expansion logic). Trailing punctuation like colons and periods are excluded from matching to avoid visual clutter in common output like "Error at ~/.bashrc:42".

gsLocker committed:
config/os: add support for clickable tilde paths

Enables clicking on tilde-based paths like ~/.config/file.txt. When
clicked, the path is expanded to the user's home directory and opened.

Related to #1972

This addresses the maintainer feedback from PR #5813 by reusing the
existing os.expandHome() function rather than duplicating tilde expansion
logic. The implementation is cross-platform (Linux/macOS/FreeBSD) from
the start.

Implementation:
- Modified URL regex to detect ~/path patterns (excludes bare ~ and ~user)
- Removed :,. from file path character class to avoid underlining punctuation
- Added negative lookbehind (?<!\.) to prevent trailing period matches
- Reuse os.expandHome() for tilde expansion in os/open.zig
- Added safety trimming for edge case punctuation handling
- Added test cases for positive and negative scenarios

The regex matches ~/path but not bare ~ or ~user/path (which would require
different expansion logic). Trailing punctuation like colons and periods
are excluded from matching to avoid visual clutter in common output like
"Error at ~/.bashrc:42".

mitchellh commented on src/os/open.zig:
```diff
@@ -20,21 +21,33 @@ pub fn open(
     kind: apprt.action.OpenUrl.Kind,
     url: []const u8,
 ) !void {
+    // Trim trailing punctuation that may have been captured by the regex
```
This is not the proper place for this. This is only a _fallback_ open helper. Each OS implements its own native opening (NSWorkspace, DBus, etc.). If this is necessary, it should be done in the URL matching itself.

jtLocker replied:
Thank you for the feedback. I moved this to the URL detection layer in Surface.zig.

mitchellh reviewed (changes requested):
I have to do a bit more review but one note.

gsLocker committed:
config/os: move tilde expansion to URL detection layer

The fallback open helper in os/open.zig is not used by all platforms. Each OS implements its own native opening mechanism (NSWorkspace on macOS, DBus on Linux, etc.), so URL processing like tilde expansion must happen before these platform handlers are called.

This moves tilde path expansion from os/open.zig to Surface.zig where URLs are initially detected and matched. Expansion now occurs in processLinks() when links are clicked and in mouseRefreshLinks() when hover previews are shown.

This ensures all OS-specific open handlers receive already-expanded paths and maintains proper separation of concerns in the URL handling pipeline.

mitchellh reviewed (changes requested):
Repeating the hostname expansion is also surely not right. Is this AI-assisted?

gsLocker committed:
config/os: add support for clickable tilde paths (#9271)

Tilde paths like ~/.config/file.txt weren't clickable before. This adds
support for Ctrl/Cmd+clicking them, expanding ~ to your home directory
when opened.

The expansion happens in the link matching layer so all the platform
handlers (NSWorkspace, DBus, etc.) get already-expanded paths. Also
added a fast path for hover/click detection that only scans the line
under the cursor instead of building a match set for the entire viewport.

Line/column numbers (like :42 or :42:10) are stripped from paths before
opening to handle compiler error output correctly.

gsLocker committed:
refactor: centralize URL preparation logic to eliminate code duplication

The tilde expansion and URL preparation logic was duplicated across 4
different functions. This refactoring consolidates it into a single
`prepareUrl()` helper function that:

- Preserves OSC8 link integrity (no trimming of authoritative links)
- Trims trailing punctuation/line numbers only from regex-detected links
- Expands tilde paths for both link types
- Returns an allocated copy with clear ownership

Changes:
- Add prepareUrl() helper that combines trimming and expansion
- Replace all expandAndDupeUrl() calls with prepareUrl()
- Fix matchSetFromOSC8Implicit to use passed prepared_url parameter
- Remove redundant expandAndDupeUrl function

@mitchellh This eliminates the code duplication you identified. However,
I notice we still perform expansion multiple times for the same URL
(e.g., once during hover preview, again on click). Should I implement
caching to avoid these redundant expansion operations, or is the current
approach acceptable?

The expansion now happens in the link matching layer as requested,
just centralized in one helper rather than scattered across functions.
